File: Algorithm/longest_consecutive_sequence.py
class Solution:
    def longestConsecutive(self, num):
        dic = {x : 0 for x in num}
        res = 0
        for x in dic:
            if not dic[x]:
                dic[x] = 1
                end = x+1
                while end in dic:
                    dic[end] = 1
                    end += 1
                if end - x > res:
                    res = end - x 
        return res              
if __name__=='__main__':
    sol=Solution()
    num=[-1,0]
    print sol.longestConsecutive(num)

File: Algorithm/CPP/longest_consecutive_sequence.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
using namespace std;


class Solution {
public:

	int longestConsecutive(vector<int>& nums) 
	{
	
		unordered_map<int,int> myMap; // use the hash map to hold the start and the end of the sequence
		int res =1;
		for (auto num : nums)
		{
			if (myMap[num]) continue;
			myMap[num] = 1;
			if (myMap.find(num -1 ) != myMap.end())
			{
				int len = myMap[num] + myMap[num - 1];
				myMap[num] = len;
				myMap[num - len + 1]++;
				res = max(res, len);
			}
			if (myMap.find(num + 1) != myMap.end())
			{
				int len = myMap[num] + myMap[num + myMap[num+1]];
				myMap[num-myMap[num]+1] = len; // the start 
				myMap[num + myMap[num + 1]] = len; // the end of the sequence
				res = max(res, len);
			}
		}
		return res;
	}

};


int main()
{
	Solution sol;
	vector<int> triangle = { 1,2,1,3,2,5};
	triangle = { 2, 1, 2, 3, 4, 1 };
	triangle = { 1, 2, 2 };
	vector<int> num = { 100, 4, 200, 1, 3, 2 };
	cout << sol.longestConsecutive(num) << endl;
	system("pause");
	return 0;
}
